sympy/polys/rings.py

```diff
@@ -3006,3 +3206,252 @@ def sqf_list(f, all=False):
+            if not common:
+                ring = polynomials[0].ring
+                domain = ring.domain
+                gcd = gcd_terms(polynomials, ring, domain)
+                return [gcd], None
+
+            syms = pi.free_variables()
+            if allsame and syms != common:
+                allsame = False
+            common &= syms
+
+        # The loop is complete if they all have the same symbols.
+        if allsame:
+            return polynomials, common
+
+        # Extract coefficients as polys containing only the common symbols.
+        all_polys = []
+        for i, pi in enumerate(polynomials):
+            coeffs_i = pi.coeff_split(pi.free_variables() - common)
+            all_polys.extend(coeffs_i)
+
+            # Quick exit:
+            if any(len(c) == 1 for c in coeffs_i):
+                ring = polynomials[0].ring
+                domain = ring.domain
+                gcd = gcd_terms((all_polys + polynomials[i+1:]), ring, domain)
+                return [gcd], None
+
+def gcd_extract(polynomials):
+    """
+    Returns the greatest common divisor (GCD) of a set of polynomials.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import gcd_extract
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> polynomials = x - y, x - y
+    >>> gcd_extract(polynomials)
+    x - y
+
+    """
+    ring = polynomials[0].ring
+    domain = ring.domain
+
+    if any(len(pi) == 1 for pi in polynomials):
+        return gcd_terms(polynomials, ring, domain)
+
+    polynomials, monom_gcd = monomial_extract(polynomials)
+
+    polynomials, common_symbols = _gcd_preprocess_polys(polynomials)
+
+    gcd = polynomials[0]
+    for pi in polynomials[1:]:
+        gcd = gcd_prs(gcd, pi)
+        if gcd == domain.one:
+            break
+
+    if monom_gcd is not None:
+        gcd = gcd * monom_gcd
+
+    return gcd
+
+def gcd_prs(p1, p2):
+    """
+    Returns the greatest common divisor (GCD) of two polynomials using the
+    Polynomial Resultant Sequences (PRS) method.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import gcd_prs
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> f = 4*x**2 + 8*x + 10
+    >>> g = 2*x**2 + 6*x + 10
+    >>> gcd_prs(f, g)
+    2
+
+    """
+    K = p1.ring.domain
+
+    x = p1.main_variable()
+
+    c1, pp1 = p1.primitive_wrt(x)
+    c2, pp2 = p2.primitive_wrt(x)
+
+    c = gcd_extract([c1, c2])
+
+    h = pp1.subresultants(pp2, x)[-1]
+    _, h = h.primitive_wrt(x)
+
+    c *= K.canonical_unit(h.LC)
+
+    h = c * h
+
+    return h
+
+def validate_prs(f, g):
+    """
+    Validates the result of the polynomial GCD computation using the ``gcd_prs``
+    algorithm with ``dmp_inner_gcd``.
+
+    """
+    ring = f.ring
+    old_gcd = ring.dmp_inner_gcd(f, g)
+    if not ring.domain.is_Field:
+        h = gcd_prs(f, g)
+        cff = f.quo(h)
+        cfg = g.quo(h)
+        new_gcd = h, cff, cfg
```

Comment:
None of the actual code/logic for computing the gcd or cofactors should be in this function because we will ultimately not want to have this function that duplicates the calculation with both the dense and sparse algorithms.